test/test_torch.py

```diff
@@ -5118,6 +5118,11 @@ def test_has_internal_overlap(self):
         b_expanded = b.expand(4, 3)
         self.assertEqual(torch._debug_has_internal_overlap(b_expanded), OVERLAP_YES)
 
+        # Check for zero strided, size 1 axis, in non-contiguous storage (gh-33812)
+        c = torch.randn(2, 1, 2).T
+        c_expanded = c.expand(2, 2, 2)[:, :1, :]
+        self.assertEqual(torch._debug_has_internal_overlap(c_expanded), OVERLAP_TOO_HARD)
```

Comment:
I'm a little confused by this test case. Which line would have failed prior to fixing this bug? Is it the `[:, :1, :]` indexing operation? I'm also curious why you're testing `OVERLAP_TOO_HARD` here.

Since it doesn't seem obvious how this test case exercises the sample code in https://github.com/pytorch/pytorch/issues/33812 ; would it perhaps also be wise to also include the sample code verbatim (maybe minimized?) as a test case?

Reply:
In this test, nothing would have failed, but _debug_has_internal_overlap would return (incorrectly) OVERLAP_YES, which would cause all in-place ops on this tensor to error out. I agree that it may be more clear to have test cases similar to failing code in #33812,

Reply:
On the other hand, since this test is part of `test_has_internal_overlap` and not a separate testcase, it makes sense to leave it as is, ignore my previous comment.

Reply:
Yeah, this is testing the root cause rather than the symptom. A minimal reproducer for the original could be:
```python
x = torch.randn(2, 1, 2).T.expand(2, 2, 2)[:, :1, :]
x.clone() # Fails here
```
Is that worth adding as its own test?

Reply:
On a higher level, I'd rather have tests that test user-visible behavior (`x.clone()`, as you suggest) than `test_has_internal_overlap` that tests implementation detail that's not user-visible, exposed specifically for testing and can change (e.g. tomorrow someone decides to stop computing memory overlap in TensorIterator, user-visible behavior will change but `test_has_internal_overlap` will happily keep passing).  But unless we decide to rewrite this whole test (I tend to think it's low priority) I think special test for this particular mode of memory overlap analysis failure is not needed. @ezyang what's your take on this?

Reply:
It might be easier to read this as the following:

```
# stride zero, size 1 axis, not contiguous
x = torch.randn(10)
y = x.as_strided([2, 1, 5], [1, 0, 2])
self.assertNotEqual(torch._debug_has_internal_overlap, OVERLAP_YES)
```

Regarding the question about what we're testing: This function, `test_has_internal_overlap`, is meant as a unit test for `has_internal_overlap`. I agree it would be nice to test the user behavior as well; my vote is to toss the user code that used to fail as a separate test case

Reply:
I agree with @ngimel. You can test both. User code is more robust to internal refactorings.